sbncode/SinglePhotonAnalysis/HelperFunctions/helper_math.h

```diff
@@ -0,0 +1,295 @@
+
+        double s1x = s1->ShowerStart().X()-vx;
+        double s1y = s1->ShowerStart().Y()-vy;
+        double s1z = s1->ShowerStart().Z()-vz;
+        double norm1  = sqrt(pow(s1x,2)+pow(s1y,2)+pow(s1z,2));
+        s1x = s1x/norm1; //unit vector pointing to shower start from point (vx, vy,vz)
+        s1y = s1y/norm1;
+        s1z = s1z/norm1;
+
+        double s2x = s2->ShowerStart().X()-vx;
+        double s2y = s2->ShowerStart().Y()-vy;
+        double s2z = s2->ShowerStart().Z()-vz;
+        double norm2  = sqrt(pow(s2x,2)+pow(s2y,2)+pow(s2z,2));
+        s2x = s2x/norm2; // unit vector pointing to shower start from point (vx, vy, vz)
+        s2y = s2y/norm2;
+        s2z = s2z/norm2;
+
+        return sqrt(2.0*E1*E2*(1.0-(s1x*s2x+s1y*s2y+s1z*s2z)));
+
+
+    }
+
+    // invariant mass of two showers, calculated directly from shower directions
+    double  invar_mass(art::Ptr<recob::Shower> & s1, double E1,  art::Ptr<recob::Shower> &s2, double E2){
+
+        double s1x = s1->Direction().X();
+        double s1y = s1->Direction().Y();
+        double s1z = s1->Direction().Z();
+
+        double s2x = s2->Direction().X();
+        double s2y = s2->Direction().Y();
+        double s2z = s2->Direction().Z();
+
+        return sqrt(2.0*E1*E2*(1.0-(s1x*s2x+s1y*s2y+s1z*s2z)));
+
+    }
+
+
+
+    // sort indices in descending order
+    template <typename T>
+        std::vector<size_t> sort_indexes(const std::vector<T> &v) {
+
+            std::vector<size_t> idx(v.size());
+            std::iota(idx.begin(), idx.end(), 0);
+
+            // sort indexes based on comparing values in v
+            std::sort(idx.begin(), idx.end(), [&v](size_t i1, size_t i2) {return v[i1] > v[i2];});
+
+            return idx;
+        }
+
+    // sort indices such that elements in v are in ascending order
+    template <typename T>
+        std::vector<size_t> sort_indexes_rev(const std::vector<T> &v) {
+
+            std::vector<size_t> idx(v.size());
+            std::iota(idx.begin(), idx.end(), 0);
+
+            // sort indexes based on comparing values in v
+            std::sort(idx.begin(), idx.end(), [&v](size_t i1, size_t i2) {return v[i1] < v[i2];});
+
+            return idx;
+        }
+
+
+    // check if two vectors have same elements (regardless of the order), and arrange their elements in order
+    template<typename T>
+        bool marks_compare_vec_nonsense(std::vector<T>& v1, std::vector<T>& v2)
+        {
+            std::sort(v1.begin(), v1.end());
+            std::sort(v2.begin(), v2.end());
+            return v1 == v2;
+        }
+
+
+
+    double calcWire(double Y, double Z, int plane, int fTPC, int fCryostat, geo::GeometryCore const& geo ){
+        double wire = geo.WireCoordinate(Y, Z, plane, fTPC, fCryostat);
+        return wire;
+    }
+
+/* unit vector orthogonal to the  wire direction of plane -- usually named as wire_dir */
+  TVector3 getWireVec(int plane){
```

Comment:
This functionality should be available through the geometry service.

Reply:
Apply the geometry service to obtain the wire angles.